core/arch/arm/kernel/spmc_sp_handler.c

```diff
@@ -52,6 +60,681 @@ static TEE_Result ffa_get_dst(struct thread_smc_args *args,
+
+	exceptions = cpu_spin_lock_xsave(&mem_ref_lock);
+	if (!receiver->ref_count) {
+		DMSG("To many relinquish requests");
+		err = FFA_DENIED;
+		goto err_unlock_memref;
+	}
+
+	receiver->ref_count--;
+	if (!receiver->ref_count) {
+		cpu_spin_unlock_xrestore(&mem_ref_lock, exceptions);
+		if (sp_unmap_ffa_regions(caller_sp, smem) != TEE_SUCCESS) {
+			DMSG("Failed to unmap region");
+			ffa_set_error(args, FFA_DENIED);
+			return;
+		}
+	} else {
+		cpu_spin_unlock_xrestore(&mem_ref_lock, exceptions);
+	}
+
+	args->a0 = FFA_SUCCESS_32;
+	return;
+
+err_unlock_rxtwx:
+	cpu_spin_unlock_xrestore(&rxtx->spinlock, exceptions);
+	ffa_set_error(args, err);
+	return;
+err_unlock_memref:
+	cpu_spin_unlock_xrestore(&mem_ref_lock, exceptions);
+	ffa_set_error(args, err);
+}
+
+static void zero_mem_region(struct sp_mem *smem, struct sp_session *s)
+{
+	void *addr = NULL;
+	struct sp_ctx *ctx = to_sp_ctx(s->ts_sess.ctx);
+	struct sp_mem_map_region *reg = NULL;
+
+	ts_push_current_session(&s->ts_sess);
+	SLIST_FOREACH(reg, &smem->regions, link) {
+		size_t sz = reg->page_count * SMALL_PAGE_SIZE;
+
+		addr = sp_mem_get_va(&ctx->uctx, reg->page_offset, reg->mobj);
+
+		assert(addr);
+		memset(addr, 0, sz);
+	}
+	ts_pop_current_session();
+}
+
+/*
+ * ffa_mem_reclaim returns false if it couldn't process the reclaim message.
+ * This happens when the memory regions was shared with the OP-TEE endpoint.
+ * After this thread_spmc calls handle_mem_reclaim() to make sure that the
+ * region is reclaimed from the OP-TEE endpoint.
+ */
+bool ffa_mem_reclaim(struct thread_smc_args *args,
+		     struct sp_session *caller_sp)
+{
+	uint64_t handle = 0;
+	uint32_t flags = args->a3;
+	uint32_t endpoint = 0;
+	struct sp_mem *smem = NULL;
+	struct sp_mem_receiver *receiver  = NULL;
+	uint32_t exceptions = 0;
+
+	handle = reg_pair_to_64(args->a2, args->a1);
+
+	smem = sp_mem_get(handle);
+	if (!smem)
+		return false;
+
+	if (caller_sp)
+		endpoint = caller_sp->endpoint_id;
+
+	/* Make sure that the caller is the owner of the share */
+	if (smem->sender_id != endpoint) {
+		ffa_set_error(args, FFA_DENIED);
+		return true;
+	}
+
+	exceptions = cpu_spin_lock_xsave(&mem_ref_lock);
+
+	/* Make sure that all shares where relinquished */
+	SLIST_FOREACH(receiver, &smem->receivers, link) {
+		if (receiver->ref_count != 0) {
+			ffa_set_error(args, FFA_DENIED);
+			cpu_spin_unlock_xrestore(&mem_ref_lock, exceptions);
+			return true;
+		}
+	}
+
+	if (flags & FFA_MEMORY_REGION_FLAG_CLEAR) {
+		if (caller_sp) {
+			zero_mem_region(smem, caller_sp);
+		} else {
+			/*
+			 * Currently we don't support zeroing Normal World
+			 * memory. To do this we would have to map the memory
+			 * again, zero it and unmap it.
+			 */
+			ffa_set_error(args, FFA_DENIED);
+			cpu_spin_unlock_xrestore(&mem_ref_lock, exceptions);
+			return true;
+		}
+	}
+
+	sp_mem_remove(smem);
+	cpu_spin_unlock_xrestore(&mem_ref_lock, exceptions);
+
+	args->a0 = FFA_SUCCESS_32;
+
+	return true;
```

Comment:
btw, `static inline bool ffa_succes(args) { args->a0 = FFA_SUCCESS_32; return true; }` is welcome.

Reply:
Done